tachyon/zk/plonk/vanishing/verifier_vanishing_argument.h

```diff
@@ -0,0 +1,120 @@
+  // Obtain a commitment to h(X) in the form of multiple pieces of degree
+  // n - 1
+  std::vector<Commitment> h_commitments;
+  size_t quotient_poly_degree = vk.constraint_system().ComputeDegree() - 1;
+  h_commitments.resize(quotient_poly_degree);
+  for (Commitment& commitment : h_commitments) {
+    if (!transcript->ReadPoint(&commitment)) return false;
+  }
+
+  *constructed_out = {std::move(h_commitments),
+                      std::move(committed).TakeRandomPolyCommitment()};
+  return true;
+}
+
+template <typename F, typename PCSTy, typename Commitment>
+[[nodiscard]] bool EvaluateAfterX(
+    VanishingConstructed<EntityTy::kVerifier, PCSTy>&& constructed,
+    crypto::TranscriptReader<Commitment>* transcript,
+    VanishingPartiallyEvaluated<PCSTy>* partially_evaluated_out) {
+  F random_eval;
+  if (!transcript->ReadScalar(&random_eval)) return false;
+
+  *partially_evaluated_out = {std::move(constructed).TakeHCommitments(),
+                              std::move(constructed).TakeRandomPolyCommitment(),
+                              std::move(random_eval)};
+  return true;
+}
+
+template <typename PCSTy, typename Evals, typename F>
+VanishingEvaluated<EntityTy::kVerifier, PCSTy> VerifyVanishingArgument(
+    VanishingPartiallyEvaluated<PCSTy>&& partially_evaluated,
+    const Evals& expressions, const crypto::Challenge255<F>& y, const F& x_n) {
+  using Commitment = typename PCSTy::Commitment;
+  using JacobianPoint =
+      typename math::internal::AdditiveSemigroupTraits<Commitment>::ReturnTy;
+
+  F y_scalar = y.ChallengeAsScalar();
+  F expected_h_eval = std::accumulate(
+      expressions.evaluations().begin(), expressions.evaluations().end(),
+      F::Zero(), [y_scalar](F& h_eval, const F& v) {
+        h_eval *= y_scalar;
+        return h_eval += v;
+      });
+  expected_h_eval /= x_n - F::One();
+
+  Commitment h_commitment =
+      std::accumulate(partially_evaluated.h_commitments().rbegin(),
+                      partially_evaluated.h_commitments().rend(),
+                      JacobianPoint::Zero(),
+                      [&x_n](JacobianPoint& acc, const Commitment& commitment) {
+                        acc = acc * x_n;
+                        return acc += commitment;
+                      })
+          .ToAffine();
+
+  return {std::move(h_commitment),
+          std::move(partially_evaluated).TakeRandomPolyCommitment(),
+          std::move(expected_h_eval),
+          std::move(partially_evaluated).TakeRandomEval()};
+}
+
+template <typename PCSTy, typename F>
+std::vector<VerifierQuery<PCSTy>> QueryVanishingArgument(
+    VanishingEvaluated<EntityTy::kVerifier, PCSTy>&& evaluated,
+    const crypto::Challenge255<F>& x) {
+  using Commitment = typename PCSTy::Commitment;
+
+  return {{x.ChallengeAsScalar(),
+           base::Ref<const Commitment>(&evaluated.h_commitment()),
+           std::move(evaluated).TakeExpectedHEval()},
+          {std::move(x.ChallengeAsScalar()),
```

Comment:
```suggestion
          {x.ChallengeAsScalar(),
```

Reply:
It's actually intended, but with my mistake.

```c++
F x_scalar = x.ChallengeAsScalar();
```
What I wanted to do was have this variable `x_scalar` and copy for the first time, then move it the second time.